### PR TITLE
VER: Yahya/5059-randomized-topology-updated

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This issue implements a randomized topology where the connectedness between every two nodes follows a fixed _edge probability_. 

Note: At this stage, the randomized topology is introduced as a _tool_ to reside on the master. However, it is not replacing the current topic-based topology yet.

### CODE REVIEW

File: network/topology/randomizedTopology.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+)
+
+// RandomizedTopology generates a random topology per channel.
+// By random topology we mean a node is connected to any other co-channel nodes with some
+// edge probability.
+type RandomizedTopology struct {
+	me      flow.Identifier             // used to keep identifier of the node
+	state   protocol.State              // used to keep a read only protocol state
+	subMngr network.SubscriptionManager // used to keep track topics the node subscribed to
+	chance  uint64                      // used to translate connectedness probability into a number in [0, 100]
```
Comment (vishalchangrani) [inline network/topology/randomizedTopology.go:21]: @yhassanzadeh13 what is the best way to interpret this? Is this the probability that two nodes would be connected?
Can I say that if node 1 has 10% probability of being down and node 2 has 20% probability of being down then `change` should be 30?
I am trying to relate it to something we know heuristically - node availability

**Review by vishalchangrani — APPROVED** [review]
thanks for the code walk through

File: network/topology/topology_utils.go [hunk 2]
```diff
@@ -20,12 +24,44 @@ func LinearFanout(size int) int {
...
+	if err != nil {
+		return nil, fmt.Errorf("could not generate hasher: %w", err)
+	}
+
+	encodedId, err := encoding.DefaultEncoder.Encode(id)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode id: %w", err)
+	}
+
+	return h.ComputeHash(encodedId), nil
```
Comment (tarakby) [inline network/topology/topology_utils.go:50]: It's a good idea to hash the ID 👌

File: network/topology/randomizedTopology.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	subMngr network.SubscriptionManager) (*RandomizedTopology, error) {
+	// edge probability should be a positive value between 0 and 1. However,
+	// we like it to be strictly greater than zero. Also, at the current scale of
+	// Flow, we need it to be greater than 0.01 to support probabilistic connectedness.
+	if edgeProb < 0.01 || edgeProb > 1 {
+		return nil, fmt.Errorf("randomized topology probability should in in range of [0.01, 1], wrong value: %f", edgeProb)
+	}
+
+	// generates seed and random number generator
+	seed, err := byteSeedFromID(nodeID)
```
Comment (tarakby) [inline network/topology/randomizedTopology.go:36]: I guess using the node ID here doesn't have any particular property compared to using a random seed (generated locally), right?

**Review by tarakby — APPROVED** [review]
Nice work Yahya 👏

File: network/topology/randomizedTopology.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/topology/randomizedTopology.go:36]: Right, however, at this point we just aim at it being deterministic. The next step would be to investigate using random beacon for sake of better entropy.

File: network/topology/topology_utils.go (hunk 2 above)
Comment (yhassanzadeh13) [reply network/topology/topology_utils.go:50]: Thanks!

File: network/topology/randomizedTopology_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+// TestUniqueness is a weak uniqueness check evaluates that no two consecutive connected components
+// on the same channel are the same.
+// It generates a topology for the first topic of consensus nodes.
+// Since topologies are seeded with the node ids, it evaluates that every two consecutive
+// topologies of the same topic for distinct nodes are distinct.
+//
+// It also checks the topology against non-inclusion of the node itself in its own topology.
+//
+// Note: currently we are using a uniform edge probability, hence there are
+// 2^(n * (n-1)) many unique topologies for the same topic across different nodes.
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/topology/randomizedTopology_test.go:82]: Exactly! nice catch! at this point, fanouts are directed edges on the topology graph.

File: network/topology/randomizedTopology.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	// samples fanout among interacting roles
+	return r.sampleFanout(ids.Filter(filter.HasRole(roles...)))
+}
+
+// tossBiasedBit returns true with probability equal `r.chance/100`, and returns false otherwise.
+func (r *RandomizedTopology) tossBiasedBit() bool {
+	draw := r.rng.UintN(100)
+	return draw <= r.chance
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/topology/randomizedTopology.go:164]: Right, but `UintN` provides a random number between `[0,N]` [exclusive](https://github.com/onflow/flow-go/blob/e47f6d79b9680b9634f966b4c95bf7fbabe21db1/crypto/random/rand.go#L7-L6) so for example if `r.chance` is `1`, having `draw < r.chance` would never be through over long-run of experiment.

File: network/topology/helper.go [hunk 6]
```diff
@@ -68,7 +60,7 @@ func CheckGraphConnected(t *testing.T, adjMap map[flow.Identifier]flow.IdentityL
 	dfs(startID, adjMap, visited, f)
 
 	// assert that expected number of nodes were visited by DFS
-	assert.Equal(t, expectedCount, len(visited))
+	require.Equal(t, expectedCount, len(visited))
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/topology/helper.go:65]: Right, added a 1-1 correspondence checking in https://github.com/onflow/flow-go/pull/253/commits/3eb90e5a159a64e14a1fc8b8c3a55662ba4ffa9f, i.e., size checking and inclusion.

File: network/topology/randomizedTopology.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/topology/randomizedTopology.go]: right, fixed in https://github.com/onflow/flow-go/pull/253/commits/1e3abbe52049f45bccbfd3d3153dc8ac293f2f1d

File: network/topology/randomizedTopology.go (hunk 3 above)
Comment (tarakby) [reply network/topology/randomizedTopology.go:36]: Thanks for the answer. I think if we don't care about the determinism anymore, even using the output of the standard package `crypto/rand` would be fine.

**Pull request merged**